# Import Libraries

In [175]:
import pandas as pd
import numpy as np
import datetime as dt

import warnings
warnings.filterwarnings('ignore') # to ignore all of the warnings when run the code

In [176]:
# to know where this file is save
import os
os.getcwd()

'/Users/salsabilafcr/Documents/DATA ANALYTICS/Projects/Python Project/Project'

# Import Dataset

In [177]:
df = pd.read_csv('/Users/salsabilafcr/Documents/DATA ANALYTICS/Projects/Python Project/Data/Online Retail Data.csv', header = 0)
df

order_id product_code                     product_name  quantity  \
0        493410      TEST001          This is a test product.         5   
1       C493411        21539          RETRO SPOTS BUTTER DISH        -1   
2        493412      TEST001          This is a test product.         5   
3        493413        21724  PANDA AND BUNNIES STICKER SHEET         1   
4        493413        84578   ELEPHANT TOY WITH BLUE T-SHIRT         1   
...         ...          ...                              ...       ...   
461768   539991        21618   4 WILDFLOWER BOTANICAL CANDLES         1   
461769   539991        72741            GRAND CHOCOLATECANDLE         4   
461770   539992        21470    FLOWER VINE RAFFIA FOOD COVER         1   
461771   539992        22258          FELT FARM ANIMAL RABBIT         1   
461772   539992        21155            RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
3       2010-01-04 09:54:00   0.85          NaN  
4       2010-01-04 09:54:00   3.75          NaN  
...                     ...    ...          ...  
461768  2010-12-23 16:49:00   1.25          NaN  
461769  2010-12-23 16:49:00   1.45          NaN  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 7 columns]

In [178]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      461773 non-null  object 
 1   product_code  461773 non-null  object 
 2   product_name  459055 non-null  object 
 3   quantity      461773 non-null  int64  
 4   order_date    461773 non-null  object 
 5   price         461773 non-null  float64
 6   customer_id   360853 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.7+ MB


# Data Cleaning

In [179]:
df_clean = df.copy()

In [180]:
# create date column
df_clean['date'] = pd.to_datetime(df_clean['order_date']).dt.date.astype('datetime64[ns]')
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      461773 non-null  object        
 1   product_code  461773 non-null  object        
 2   product_name  459055 non-null  object        
 3   quantity      461773 non-null  int64         
 4   order_date    461773 non-null  object        
 5   price         461773 non-null  float64       
 6   customer_id   360853 non-null  float64       
 7   date          461773 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 28.2+ MB


In [181]:
# remove null value in customer_id column
df_clean = df_clean[~df_clean['customer_id'].isna()]
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 360853 entries, 0 to 461744
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      360853 non-null  object        
 1   product_code  360853 non-null  object        
 2   product_name  360853 non-null  object        
 3   quantity      360853 non-null  int64         
 4   order_date    360853 non-null  object        
 5   price         360853 non-null  float64       
 6   customer_id   360853 non-null  float64       
 7   date          360853 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 24.8+ MB


In [182]:
# makes all product_name columns lowercase
df_clean['product_name'] = df_clean['product_name'].str.lower()
df_clean.head(3)

order_id product_code             product_name  quantity  \
0   493410      TEST001  this is a test product.         5   
1  C493411        21539  retro spots butter dish        -1   
2   493412      TEST001  this is a test product.         5   

            order_date  price  customer_id       date  
0  2010-01-04 09:24:00   4.50      12346.0 2010-01-04  
1  2010-01-04 09:43:00   4.25      14590.0 2010-01-04  
2  2010-01-04 09:53:00   4.50      12346.0 2010-01-04

In [183]:
# remove all rows with product_code or product_name 'test'
df_clean = df_clean[~(df_clean['product_code'].str.lower().str.contains('test')) | 
                    ~(df_clean['product_name'].str.lower().str.contains('test'))]

df_clean.head()

order_id product_code                     product_name  quantity  \
1  C493411        21539          retro spots butter dish        -1   
6   493414        21844                   retro spot mug        36   
7   493414        21533        retro spot large milk jug        12   
8   493414        37508  new england ceramic cake server         2   
9   493414       35001G             hand open shape gold         2   

            order_date  price  customer_id       date  
1  2010-01-04 09:43:00   4.25      14590.0 2010-01-04  
6  2010-01-04 10:28:00   2.55      14590.0 2010-01-04  
7  2010-01-04 10:28:00   4.25      14590.0 2010-01-04  
8  2010-01-04 10:28:00   2.55      14590.0 2010-01-04  
9  2010-01-04 10:28:00   4.25      14590.0 2010-01-04

In [184]:
# create an order_status column with the value 'canceled' if the order_id begins with the letter 'c' and 
# 'delivered' if the order_id does not begin with the letter 'c'
df_clean['order_status'] = np.where(df_clean['order_id'].str[:1] == 'C', 'cancelled', 'delivered')
df_clean.head(3)

order_id product_code               product_name  quantity  \
1  C493411        21539    retro spots butter dish        -1   
6   493414        21844             retro spot mug        36   
7   493414        21533  retro spot large milk jug        12   

            order_date  price  customer_id       date order_status  
1  2010-01-04 09:43:00   4.25      14590.0 2010-01-04    cancelled  
6  2010-01-04 10:28:00   2.55      14590.0 2010-01-04    delivered  
7  2010-01-04 10:28:00   4.25      14590.0 2010-01-04    delivered

In [185]:
# change the negative quantity value to positive because the negative value only indicates 
# that the order is canceled
df_clean['quantity'] = df_clean['quantity'].abs()
df_clean.head(3)

order_id product_code               product_name  quantity  \
1  C493411        21539    retro spots butter dish         1   
6   493414        21844             retro spot mug        36   
7   493414        21533  retro spot large milk jug        12   

            order_date  price  customer_id       date order_status  
1  2010-01-04 09:43:00   4.25      14590.0 2010-01-04    cancelled  
6  2010-01-04 10:28:00   2.55      14590.0 2010-01-04    delivered  
7  2010-01-04 10:28:00   4.25      14590.0 2010-01-04    delivered

In [186]:
# creates an amount value column, which is the multiplication of quantity and price
df_clean['amount'] = df_clean['quantity'] * df_clean['price']
df_clean.head(3)

order_id product_code               product_name  quantity  \
1  C493411        21539    retro spots butter dish         1   
6   493414        21844             retro spot mug        36   
7   493414        21533  retro spot large milk jug        12   

            order_date  price  customer_id       date order_status  amount  
1  2010-01-04 09:43:00   4.25      14590.0 2010-01-04    cancelled    4.25  
6  2010-01-04 10:28:00   2.55      14590.0 2010-01-04    delivered   91.80  
7  2010-01-04 10:28:00   4.25      14590.0 2010-01-04    delivered   51.00

In [187]:
# replace the product_name of a product_code that has multiple product_names 
# with one of its most frequently occurring product_names
most_freq_product_name = df_clean.groupby(['product_code','product_name'], as_index=False).agg(order_cnt=('order_id','nunique')).sort_values(['product_code','order_cnt'], ascending=[True,False])
most_freq_product_name['rank'] = most_freq_product_name.groupby('product_code')['order_cnt'].rank(method='first', ascending=False)
most_freq_product_name = most_freq_product_name[most_freq_product_name['rank']==1].drop(columns=['order_cnt','rank'])
df_clean = df_clean.merge(most_freq_product_name.rename(columns={'product_name':'most_freq_product_name'}), how='left', on='product_code')
df_clean['product_name'] = df_clean['most_freq_product_name']
df_clean = df_clean.drop(columns='most_freq_product_name')
df_clean.head(3)

order_id product_code               product_name  quantity  \
0  C493411        21539  red retrospot butter dish         1   
1   493414        21844          red retrospot mug        36   
2   493414        21533  retro spot large milk jug        12   

            order_date  price  customer_id       date order_status  amount  
0  2010-01-04 09:43:00   4.25      14590.0 2010-01-04    cancelled    4.25  
1  2010-01-04 10:28:00   2.55      14590.0 2010-01-04    delivered   91.80  
2  2010-01-04 10:28:00   4.25      14590.0 2010-01-04    delivered   51.00

In [188]:
# convert customer_id column to string
df_clean['customer_id'] = df_clean['customer_id'].astype(str)
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 360842 entries, 0 to 360841
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      360842 non-null  object        
 1   product_code  360842 non-null  object        
 2   product_name  360842 non-null  object        
 3   quantity      360842 non-null  int64         
 4   order_date    360842 non-null  object        
 5   price         360842 non-null  float64       
 6   customer_id   360842 non-null  object        
 7   date          360842 non-null  datetime64[ns]
 8   order_status  360842 non-null  object        
 9   amount        360842 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(6)
memory usage: 27.5+ MB


In [189]:
# remove outliers
from scipy import stats
df_clean = df_clean[(np.abs(stats.zscore(df_clean[['quantity','amount']]))<3).all(axis=1)]
df_clean = df_clean.reset_index(drop=True)
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358482 entries, 0 to 358481
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      358482 non-null  object        
 1   product_code  358482 non-null  object        
 2   product_name  358482 non-null  object        
 3   quantity      358482 non-null  int64         
 4   order_date    358482 non-null  object        
 5   price         358482 non-null  float64       
 6   customer_id   358482 non-null  object        
 7   date          358482 non-null  datetime64[ns]
 8   order_status  358482 non-null  object        
 9   amount        358482 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(6)
memory usage: 27.4+ MB


# Create RFM Segmentation

In [190]:
df_user = df_clean.groupby('customer_id', as_index = False).agg(order_cnt = ('order_id','nunique'),
                                                                max_order_date = ('date', 'max'),
                                                                total_order_value = ('amount', 'sum'))
df_user

customer_id  order_cnt max_order_date  total_order_value
0        12346.0          5     2010-10-04             602.40
1        12608.0          1     2010-10-31             415.79
2        12745.0          2     2010-08-10             723.85
3        12746.0          2     2010-06-30             266.35
4        12747.0         19     2010-12-13            4094.79
...          ...        ...            ...                ...
3884     18283.0          6     2010-11-22             641.77
3885     18284.0          2     2010-10-06             486.68
3886     18285.0          1     2010-02-17             427.00
3887     18286.0          2     2010-08-20             941.48
3888     18287.0          4     2010-11-22            2345.71

[3889 rows x 4 columns]

Aggregate transaction data to create RFM columns, namely the last order date of each user (max_order_date), total transactions (order_cnt), and total transaction value (total_order_value).

In [191]:
today = df_clean['date'].max()
df_user['day_since_last_order'] = (today - df_user['max_order_date']).dt.days
df_user.head()

customer_id  order_cnt max_order_date  total_order_value  \
0     12346.0          5     2010-10-04             602.40   
1     12608.0          1     2010-10-31             415.79   
2     12745.0          2     2010-08-10             723.85   
3     12746.0          2     2010-06-30             266.35   
4     12747.0         19     2010-12-13            4094.79   

   day_since_last_order  
0                    80  
1                    53  
2                   135  
3                   176  
4                    10

In [192]:
today

Timestamp('2010-12-23 00:00:00')

Create a column for the number of days since the customer placed their last order. We assume 'today' is the date of the last order in the dataset.

In [193]:
df_user.describe()

order_cnt                 max_order_date  total_order_value  \
count  3889.000000                           3889        3889.000000   
mean      5.128568  2010-09-23 04:46:57.793777664        1544.623084   
min       1.000000            2010-01-05 00:00:00           1.250000   
25%       1.000000            2010-08-19 00:00:00         296.360000   
50%       3.000000            2010-10-26 00:00:00         648.200000   
75%       6.000000            2010-11-28 00:00:00        1585.940000   
max     163.000000            2010-12-23 00:00:00       71970.390000   
std       8.499330                            NaN        3434.816315   

       day_since_last_order  
count           3889.000000  
mean              90.800720  
min                0.000000  
25%               25.000000  
50%               58.000000  
75%              126.000000  
max              352.000000  
std               88.873286

In [194]:
df_user['recency_score'] = pd.cut(df_user['day_since_last_order'],
                                 bins = [df_user['day_since_last_order'].min(),
                                        np.percentile(df_user['day_since_last_order'], 20),
                                        np.percentile(df_user['day_since_last_order'], 40),
                                        np.percentile(df_user['day_since_last_order'], 60),
                                        np.percentile(df_user['day_since_last_order'], 80),
                                        df_user['day_since_last_order'].max()],
                                 labels = [5, 4, 3, 2, 1],
                                 include_lowest = True).astype(int)
df_user.head()

customer_id  order_cnt max_order_date  total_order_value  \
0     12346.0          5     2010-10-04             602.40   
1     12608.0          1     2010-10-31             415.79   
2     12745.0          2     2010-08-10             723.85   
3     12746.0          2     2010-06-30             266.35   
4     12747.0         19     2010-12-13            4094.79   

   day_since_last_order  recency_score  
0                    80              2  
1                    53              3  
2                   135              2  
3                   176              1  
4                    10              5

Create a binning of the number of days since the last order consisting of 5 bins with boundaries, namely, Min, P20, P40, P60, P80, Max. Then, label 1 to 5 from the highest bin to the lowest as the recency score.

In [195]:
df_user['frequency_score'] = pd.cut(df_user['order_cnt'],
                                 bins = [ 0, # df_user['order_cnt'].min(),<-- cannot be used because 
                                             # from the descriptive statistics the value of the 20th percentile 
                                             # (order_cnt) is only 1, so the lower limit cannot be 1 as well.
                                        np.percentile(df_user['order_cnt'], 20),
                                        np.percentile(df_user['order_cnt'], 40),
                                        np.percentile(df_user['order_cnt'], 60),
                                        np.percentile(df_user['order_cnt'], 80),
                                        df_user['order_cnt'].max()],
                                 labels = [1, 2, 3, 4, 5],
                                 include_lowest = True).astype(int)
df_user.head()

customer_id  order_cnt max_order_date  total_order_value  \
0     12346.0          5     2010-10-04             602.40   
1     12608.0          1     2010-10-31             415.79   
2     12745.0          2     2010-08-10             723.85   
3     12746.0          2     2010-06-30             266.35   
4     12747.0         19     2010-12-13            4094.79   

   day_since_last_order  recency_score  frequency_score  
0                    80              2                4  
1                    53              3                1  
2                   135              2                2  
3                   176              1                2  
4                    10              5                5

Create a binning of the total transactions (orders) consisting of 5 bins with boundaries, namely, Min, P20, P40, P60, P80, Max. Then, label 1 to 5 from the lowest bin to the highest as the frequency score.

In [196]:
df_user['monetary_score'] = pd.cut(df_user['total_order_value'],
                                 bins = [df_user['total_order_value'].min(),
                                        np.percentile(df_user['total_order_value'], 20),
                                        np.percentile(df_user['total_order_value'], 40),
                                        np.percentile(df_user['total_order_value'], 60),
                                        np.percentile(df_user['total_order_value'], 80),
                                        df_user['total_order_value'].max()],
                                 labels = [1, 2, 3, 4, 5],
                                 include_lowest = True).astype(int)
df_user.head()

customer_id  order_cnt max_order_date  total_order_value  \
0     12346.0          5     2010-10-04             602.40   
1     12608.0          1     2010-10-31             415.79   
2     12745.0          2     2010-08-10             723.85   
3     12746.0          2     2010-06-30             266.35   
4     12747.0         19     2010-12-13            4094.79   

   day_since_last_order  recency_score  frequency_score  monetary_score  
0                    80              2                4               3  
1                    53              3                1               2  
2                   135              2                2               3  
3                   176              1                2               2  
4                    10              5                5               5

Create a binning of the total order value consisting of 5 bins with boundaries, namely, Min, P20, P40, P60, P80, Max. Then, label 1 to 5 from the lowest bin to the highest as the monetery score.

In [197]:
df_user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3889 entries, 0 to 3888
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   customer_id           3889 non-null   object        
 1   order_cnt             3889 non-null   int64         
 2   max_order_date        3889 non-null   datetime64[ns]
 3   total_order_value     3889 non-null   float64       
 4   day_since_last_order  3889 non-null   int64         
 5   recency_score         3889 non-null   int64         
 6   frequency_score       3889 non-null   int64         
 7   monetary_score        3889 non-null   int64         
dtypes: datetime64[ns](1), float64(1), int64(5), object(1)
memory usage: 243.2+ KB


In [202]:
df_user['segment'] = np.select(
       [(df_user['recency_score'] == 5) & (df_user['frequency_score'] >= 4),
        (df_user['recency_score'].between(3, 4)) & (df_user['frequency_score'] >= 4),
        (df_user['recency_score'] >= 4) & (df_user['frequency_score'].between(2, 3)),
        (df_user['recency_score'] <= 2) & (df_user['frequency_score'] == 5),
        (df_user['recency_score'] == 3) & (df_user['frequency_score'] == 3),
        (df_user['recency_score'] == 5) & (df_user['frequency_score'] == 1),
        (df_user['recency_score'] == 4) & (df_user['frequency_score'] ==1),
        (df_user['recency_score'] <=2) & (df_user['frequency_score'].between(3, 4)),
        (df_user['recency_score'] == 3) & (df_user['frequency_score'] <= 2),
        (df_user['recency_score'] <= 2) & (df_user['frequency_score'] <= 2)],
       ['01-Champion', '02-Loyal Customers', '03-Potential Loyalists', "04-Can't Lose Them", '05-No Need Attention',
       '06-New Customers', '07-Promising', '08-At Risk', '09-About to Sleep', '10-Hibernating'])
df_user

customer_id  order_cnt max_order_date  total_order_value  \
0        12346.0          5     2010-10-04             602.40   
1        12608.0          1     2010-10-31             415.79   
2        12745.0          2     2010-08-10             723.85   
3        12746.0          2     2010-06-30             266.35   
4        12747.0         19     2010-12-13            4094.79   
...          ...        ...            ...                ...   
3884     18283.0          6     2010-11-22             641.77   
3885     18284.0          2     2010-10-06             486.68   
3886     18285.0          1     2010-02-17             427.00   
3887     18286.0          2     2010-08-20             941.48   
3888     18287.0          4     2010-11-22            2345.71   

      day_since_last_order  recency_score  frequency_score  monetary_score  \
0                       80              2                4               3   
1                       53              3                1               2   
2                      135              2                2               3   
3                      176              1                2               2   
4                       10              5                5               5   
...                    ...            ...              ...             ...   
3884                    31              4                4               3   
3885                    78              2                2               3   
3886                   309              1                1               2   
3887                   125              2                2               4   
3888                    31              4                3               5   

                    segment  
0                08-At Risk  
1         09-About to Sleep  
2            10-Hibernating  
3            10-Hibernating  
4               01-Champion  
...                     ...  
3884     02-Loyal Customers  
3885         10-Hibernating  
3886         10-Hibernating  
3887         10-Hibernating  
3888  03-Potential Loyalist  

[3889 rows x 9 columns]

Create segment columns based on the recency and frequency scores.

In [208]:
summary = pd.pivot_table(df_user, index = 'segment',
              values = ['customer_id', 'day_since_last_order', 'order_cnt', 'total_order_value'],
              aggfunc = {'customer_id' : pd.Series.nunique,
                        'day_since_last_order' : [np.mean, np.median],
                        'order_cnt' : [np.mean, np.median],
                        'total_order_value' : [np.mean, np.median]})
summary['pct_unique'] = (summary['customer_id'] / summary['customer_id'].sum() * 100).round(1) # calculating user percentage
summary

customer_id day_since_last_order         order_cnt  \
                          nunique                 mean median       mean   
segment                                                                    
01-Champion                   550            10.618182    9.5  15.467273   
02-Loyal Customers            546            40.864469   37.0   8.767399   
03-Potential Loyalist         523            23.573614   24.0   2.829828   
04-Can't Lose Them             64           121.984375  112.5  11.375000   
05-No Need Attention          176            58.613636   59.0   3.397727   
06-New Customers               50            14.220000   16.0   1.000000   
07-Promising                  142            32.760563   34.0   1.000000   
08-At Risk                    426           140.455399  120.0   4.136150   
09-About to Sleep             352            58.735795   58.0   1.417614   
10-Hibernating               1060           196.837736  199.0   1.313208   

                             total_order_value           pct_unique  
                      median              mean    median             
segment                                                              
01-Champion             10.0       5003.674245  2775.525       14.1  
02-Loyal Customers       7.0       2622.817826  1946.850       14.0  
03-Potential Loyalist    3.0        766.769828   622.070       13.4  
04-Can't Lose Them       9.5       2839.948125  2268.405        1.6  
05-No Need Attention     3.0        989.232676   826.370        4.5  
06-New Customers         1.0        244.689000   193.675        1.3  
07-Promising             1.0        287.800282   238.440        3.7  
08-At Risk               4.0       1153.825683   875.430       11.0  
09-About to Sleep        1.0        448.229688   334.755        9.1  
10-Hibernating           1.0        343.083842   257.005       27.3

# Analysis Result

From the results of customer segmentation using the RFM method, there are several insights, including: 
1. Most users are in the Hibernating segment (1060 or 27.3% customers), Champion (550 or 14.1% customers), Loyal Customers (546 or 14% customers).
2. Special programs that focus on the urgency of transactions for Loyal Customers who have an average last transaction of 40 days ago, thus making them transact again in the near future and can move up to the Champion segment.
3. Special programs that focus on the number of transactions for Potential Loyalists who have an average number of orders of 2 times, thus making them transact more frequently and can move up to the Champion segment.
4. Special programs for the Hibernating segment can be carried out to make them return to transactions even though they are not so frequent, thus they can move up to the New Customers or even Potential Loyalists segment.